### PR TITLE
570 forecasting rdbnamespace per request

### PR DESCRIPTION
**Pull request by markushofmeister**
This PR adds the following functionality to the Forecasting Agent (as of version 1.1.0):
- connection properties (i.e. `namespace`, `database`, `db_url`, ... ) provided in the `docker-compose` files only represent default values
- those default values can be overwritten on a per request basis by respective keys in the HTTP request to the agent

The functionality has been tested successfully locally outside the stack.

### CODE REVIEW

File: Agents/ForecastingAgent/forecasting/utils/default_configs.py [hunk 1]
```diff
@@ -42,45 +48,46 @@ def retrieve_settings():
...
-            raise ValueError('No "DATABASE" value has been provided in environment variables.')
-        if DATABASE != 'postgres':
+            warning_msg = f'No default "DATABASE" value has been provided in environment variables. '
+            warning_msg += f'Ensure that "DATABASE" is provided in HTTP request to the agent to avoid issues.'
+            logger.warning(warning_msg)
+        elif DATABASE != 'postgres':
             logger.warning(f'Provided "DATABASE" name {DATABASE} does not match default database name "postgres".')
         
         # Retrieve settings from Stack Clients
+        # TODO: Review and check what happens/is returned if empty strings are provided
```
Comment (markushofmeister) [inline Agents/ForecastingAgent/forecasting/utils/default_configs.py:90]: This part requires testing for stack deployment, i.e. What happens when no default values are provided? In case exceptions are thrown this would require rework

File: Agents/ForecastingAgent/forecasting/flaskapp/forecastingtasks/routes.py [hunk 2]
```diff
@@ -97,3 +122,53 @@ def api_forecast():
...
+                # ... otherwise use default value
+                default = eval(r.upper())
+                if default == '':                    
+                    logger.error(f'No "{r}" value provided in HTTP request despite missing default value.')
+                    raise InvalidInput(f'No "{r}" value provided in HTTP request, despite missing default value.')
+                else:
+                    stack_params[r] = default
+        # Retrieve connection parameters from stack clients (returned as:
+        # DB_URL, DB_USER, DB_PASSWORD, QUERY_ENDPOINT, UPDATE_ENDPOINT)
+        stack_conn_param = retrieve_stack_settings(**stack_params)
```
Comment (markushofmeister) [inline Agents/ForecastingAgent/forecasting/flaskapp/forecastingtasks/routes.py:168]: This should also be verified in stack deployment

File: Agents/ForecastingAgent/stack-manager-input-config/forecasting-agent.json [hunk 3]
```diff
@@ -38,7 +38,7 @@
     },
     "endpoints": {
         "rest": {
-            "url": "http://localhost:5000/api/forecastingAgent/",
+            "url": "http://localhost:5005/api/forecastingAgent/",
```
Comment (sm453) [inline Agents/ForecastingAgent/stack-manager-input-config/forecasting-agent.json:41]: Please undo this change. It should stay at 5000. NB This is the internal number, not the external one that you changed in the docker compose files.

File: Agents/ForecastingAgent/forecasting/utils/default_configs.py (hunk 1 above)
Comment (sm453) [reply Agents/ForecastingAgent/forecasting/utils/default_configs.py:90]: > This part requires testing for stack deployment, i.e. What happens when no default values are provided? In case exceptions are thrown this would require rework

If no namespace environment variable is defined in the agent config json, then the container fails to start when attempting to do so with the stack manager. The container logs show `ERROR - "NAMESPACE" name is missing in environment variables`. I am happy with this behaviour, and the error message is sufficiently clear.

**Review by sm453 — CHANGES_REQUESTED** [review]
I have successfully managed to produce a time series forecast within a stack, with a database and namespace passed in an http request (that differ from the defaults set by environment variables).

The only change required in my view is to revert back to port 5000 in the agent config json (and only there - not in the docker compose files!).

**Comment by markushofmeister** [comment]
Updated agent behavior:
1) All environment variables in `docker-compose.yml` except `STACK_NAME` are now optional: If provided they serve as default value (to be potentially overwritten by equivalent in HTTP request), otherwise no default is set and key is required in HTTP request (which is optional otherwise)
2) Updated README to 1) reflect that updated behavior and 2) include section about potentially troubleshoot nginx routing issue
3) Aligned internal agent routes to ease stack deployment using `.json` file, i.e. forecasting route and startup route follow same pattern now

Docker images have been pushed to ghcr.io

@sm453: Do we still need the house45 related files in here? Otherwise, we might want to remove them. Just placed them here initially as examples

**Comment by kokfoongcmcl** [comment]
FYI I'm adding the option to specify the time series schema in the time series client, so you might not need to have multiple databases

**Comment by sm453** [comment]
> FYI I'm adding the option to specify the time series schema in the time series client, so you might not need to have multiple databases

We have many entities (agents) that do many things with many different time series. It is not reasonable to assume that all time series are in the same schema/database - and they shouldn't be, for a number of reasons. Therefore, all entities (agents) _must_ offer the functionality to be passed which schema/database to use for a particular purpose.

**Comment by sm453** [comment]
> @sm453: Do we still need the house45 related files in here? Otherwise, we might want to remove them. Just placed them here initially as examples

Just for the record, yes, please. I have found them useful as examples how it can be used.
